### ISSUE TITLE
Query about Latency Calculation Scripts (FT, LT, and Avg. Latency)

### ISSUE THREAD

**Issue by xiaodaien** [Audio-WestlakeU/UMA-ASR#4]
Hi ！I am reproducing your paper on AISHELL-1 and AISHELL-2. I attempted to write my own scripts for calculating FT, LT, and Avg. latency, but I’m getting negative values, likely due to timestamp alignment issues.Could you please share the original scripts used to extract output timestamps and compare them with MFA ground-truth boundaries?
Thank you for your great work and for open-sourcing the project!
                                                                                              Best regards

**Comment by FnoY0723** [comment]
1. load textgrid of MFA in espnet2/bin/asr_unimodal_inference.py
 ```
if self.calculate_latency:
     import textgrid
     if 'BAC' in utt_id:
         text_grid_path = f'aishell/asr1/downloads/forced_alignment/test/{utt_id[6:11]}/{utt_id}.TextGrid'
     elif 'AT' in utt_id:
         text_grid_path = f'aishell2/AISHELL-DEV-TEST-SET/Android/test/forced_alignments/{utt_id[1:6]}/{utt_id}.TextGrid'
     elif 'IT' in utt_id:
         text_grid_path = f'aishell2/AISHELL-DEV-TEST-SET/iOS/test/forced_alignments/{utt_id[1:6]}/{utt_id}.TextGrid'
     elif 'MT' in utt_id:
         text_grid_path = f'aishell2/AISHELL-DEV-TEST-SET/Mic/test/forced_alignments/{utt_id[1:6]}/{utt_id}.TextGrid'
     else:
         text_grid_path = None
    
     logging.info(f"text_grid_path: {text_grid_path}")
     utt_tg = textgrid.TextGrid.fromFile(text_grid_path)[0]
```
2. return `valley_index_end` from UMA forward  without early termination in https://github.com/Audio-WestlakeU/UMA-ASR/blob/f854cb37dc6b8cd67cb6fa58b904c617796ace80/espnet2/asr/uma.py#L113
```
return xs_pad, olens, valley_index_end
```
3. obtain the timestamp predicted by UMA in https://github.com/Audio-WestlakeU/UMA-ASR/blob/f854cb37dc6b8cd67cb6fa58b904c617796ace80/espnet2/bin/asr_unimodal_inference.py#L331
```
enc, umalen, scalar_importance = self.asr_model.uma(enc, enclen)
enc, _ = self.asr_model.decoder(enc, umalen, torch.tensor(0), torch.tensor(0), self.asr_model.ctc)
# Normal ASR
if isinstance(enc, tuple):
    enc = enc[0]

# obtain the tokens corresponding to the UMA output frames
ys_hat = (self.asr_model.ctc.argmax(enc).data)[0].tolist()
token = self.converter.ids2tokens(ys_hat)

valleys = scalar_importance.squeeze().cpu()
valleys = valleys.tolist()
utt_latency = {}
for interval in utt_tg:
    start_time = interval.minTime
    end_time = interval.maxTime
    text = interval.mark
    if text in token:
        # for simplicity, when there are repeated tokens, only the latency of the first occurrence will be counted.
        index = token.index(text)
        reg_time = (valleys[index]-1)*32
        # if use lookahead: latency = reg_time - end_time*1000 + lookahead
        latency = reg_time - end_time*1000
           
        if  start_time*1000 <= reg_time:
            all_latency.append(latency)
            utt_latency[text] = int(latency)
            logging.info(f'latency: {text} {int(latency)}ms')
```

**Comment by xiaodaien** [comment]
Hi @FnoY0723 Ying Fang

   Thank you for sharing the code and the impressive work on UMA-ASR.I am currently working on reproducing the results using the provided conf/train_asr_uma_mamba.yaml. I noticed a detail in the latency calculation logic in asr_unimodal_inference.py that I would like to clarify.In the inference code, the frame duration is hardcoded as 32ms:
`reg_time = (valleys[index]-1)*32`
However, according to the train_asr_uma_mamba.yaml config:
   The frontend is set to default (which usually implies hop_length=160 for 16k fs, i.e., 10ms).
   The encoder input layer is causal_conv2d (which typically performs 4x subsampling).
Theoretically, the frame shift at the encoder output should be 10ms *  4 = 40ms.
Could you kindly clarify if the 32ms was intended for a specific frontend configuration (e.g., 8ms hop size) that might be different from the provided config? Or should I use 40ms when using the default configuration?
I am asking this to ensure I can reproduce the baseline results (First Token / Last Token Latency) as accurately as possible.
Thanks for your help!

**Comment by FnoY0723** [comment]
> Hi [@FnoY0723](https://github.com/FnoY0723) Ying Fang
> 
> Thank you for sharing the code and the impressive work on UMA-ASR.I am currently working on reproducing the results using the provided conf/train_asr_uma_mamba.yaml. I noticed a detail in the latency calculation logic in asr_unimodal_inference.py that I would like to clarify.In the inference code, the frame duration is hardcoded as 32ms: `reg_time = (valleys[index]-1)*32` However, according to the train_asr_uma_mamba.yaml config: The frontend is set to default (which usually implies hop_length=160 for 16k fs, i.e., 10ms). The encoder input layer is causal_conv2d (which typically performs 4x subsampling). Theoretically, the frame shift at the encoder output should be 10ms * 4 = 40ms. Could you kindly clarify if the 32ms was intended for a specific frontend configuration (e.g., 8ms hop size) that might be different from the provided config? Or should I use 40ms when using the default configuration? I am asking this to ensure I can reproduce the baseline results (First Token / Last Token Latency) as accurately as possible. Thanks for your help!

32ms indeed refers to the model where the STFT hop size is set to 8ms. Note that in ESPNet toolkit, the default settings for STFT (window length, hop size) are (512, 128). https://github.com/espnet/espnet/blob/72a3db47bb26bf3ac2d43b055b517572bce67e38/espnet2/asr/frontend/default.py#L29

**Comment by xiaodaien** [comment]
Hi @FnoY0723 ,

Thank you for your impressive work on UMA-ASR. I am currently trying to reproduce your results on the AISHELL-1 dataset using the provided configuration (conf/train_asr_uma_mamba.yaml), but I have encountered two major issues that I hope you could shed some light on.
1. Training Instability (Model Collapse)
I am training the model using NVIDIA RTX 4090 GPUs. Using the default configuration (lr=0.001, grad_clip=5.0, amp=true), I observed severe model collapse where the loss spikes and CER increases drastically after a certain epoch.
My Training Curve (Collapse):
<img width="474" height="375" alt="Image" src="https://github.com/user-attachments/assets/bba8a47c-92fc-46d6-81fe-0e2680096948" />
Reference Curve (from your log)
<img width="590" height="447" alt="Image" src="https://github.com/user-attachments/assets/b57693aa-768b-4a04-be73-86205ef8687d" />
Could this be related to the mixed-precision (AMP) stability on 4090 GPUs? Or do you suggest adjusting the learning rate/gradient clipping for this specific hardware environment?
2. Latency Calculation Logic
 I am also trying to calculate the FT (First Token), LT (Last Token), and Avg latency to match the metrics reported in your paper. Based on the code analysis and our previous discussions (regarding the 32ms frame shift), I implemented the calculation logic as follows.
 However, I am finding it difficult to reproduce the exact metrics. My main uncertainty lies in how repeated tokens are handled.
Here is my inference code snippet:
`# ... (Encoder Forward) ...
enc, umalen, valley_index_end = self.asr_model.uma(enc, enclen)

if utt_id is not None:
    # ... (MFA path setup) ...
    FRAME_DURATION = 32  # Using 32ms as confirmed
    
    # ... (Load TextGrid and ASR Tokens) ...
    
    for interval in tg:
        text = interval.mark
        if not text: continue

        target_char = text[-1] # Target character from MFA

        if target_char in token_list:
            # My current logic: finds the FIRST occurrence of the character
            index = token_list.index(target_char) 

            if index < len(valleys):
                reg_time = (valleys[index] - 1) * FRAME_DURATION
                
                # Calculate latency
                latency = reg_time - interval.maxTime * 1000
                all_latency.append(latency)`
My Questions:
1.Handling Repeated Tokens: My code uses token_list.index(target_char), which always returns the index of the first occurrence of the character. For sentences with repeated characters (e.g., "人人..."), this logic results in very low (or negative) latency for the second occurrence. Is this "First Match" logic the one used in your paper's baseline? Or should I implement a cursor/pointer to match characters sequentially?

2.Filter Condition: Do you apply any filtering (e.g., if prediction_time < start_time) to exclude these "negative latency" cases, or do you include all matched tokens in the final average?

Any advice or the original inference script for metric calculation would be greatly appreciated!

Thanks!

**Comment by FnoY0723** [comment]
1. To stable training
       During development, I found that Mamba tends to collapse more frequently during speech sequence tasks. However, it can be handled by using weight_decay. The specific optimizer settings are as follows:
```
optim: adamw
optim_conf:
   lr: 0.001
   weight_decay: 0.01
```
2. I have already replied regarding the code for calculating latency. 
      Considering that the frequency of repeated words is very low, and latency is a statistical variable, for simplicity, only the latency of the first occurrence will be counted.

> 1. load textgrid of MFA in espnet2/bin/asr_unimodal_inference.py
> 
> ```
> if self.calculate_latency:
>     import textgrid
>     if 'BAC' in utt_id:
>         text_grid_path = f'aishell/asr1/downloads/forced_alignment/test/{utt_id[6:11]}/{utt_id}.TextGrid'
>     elif 'AT' in utt_id:
>         text_grid_path = f'aishell2/AISHELL-DEV-TEST-SET/Android/test/forced_alignments/{utt_id[1:6]}/{utt_id}.TextGrid'
>     elif 'IT' in utt_id:
>         text_grid_path = f'aishell2/AISHELL-DEV-TEST-SET/iOS/test/forced_alignments/{utt_id[1:6]}/{utt_id}.TextGrid'
>     elif 'MT' in utt_id:
>         text_grid_path = f'aishell2/AISHELL-DEV-TEST-SET/Mic/test/forced_alignments/{utt_id[1:6]}/{utt_id}.TextGrid'
>     else:
>         text_grid_path = None
>    
>     logging.info(f"text_grid_path: {text_grid_path}")
>     utt_tg = textgrid.TextGrid.fromFile(text_grid_path)[0]
> ```
> 
> 2. return `valley_index_end` from UMA forward  without early termination in 
>      
>        
>          [UMA-ASR/espnet2/asr/uma.py](https://github.com/Audio-WestlakeU/UMA-ASR/blob/f854cb37dc6b8cd67cb6fa58b904c617796ace80/espnet2/asr/uma.py#L113)
>        
>        
>             Line 113
>          in
>          [f854cb3](/Audio-WestlakeU/UMA-ASR/commit/f854cb37dc6b8cd67cb6fa58b904c617796ace80)
>        
>      
>      
>        
>    
>            
>              
>               return xs_pad, olens, None
> 
> ```
> return xs_pad, olens, valley_index_end
> ```
> 
> 3. obtain the timestamp predicted by UMA in 
>      
>        
>          [UMA-ASR/espnet2/bin/asr_unimodal_inference.py](https://github.com/Audio-WestlakeU/UMA-ASR/blob/f854cb37dc6b8cd67cb6fa58b904c617796ace80/espnet2/bin/asr_unimodal_inference.py#L331)
>        
>        
>             Line 331
>          in
>          [f854cb3](/Audio-WestlakeU/UMA-ASR/commit/f854cb37dc6b8cd67cb6fa58b904c617796ace80)
>        
>      
>      
>        
>    
>            
>              
>               enc, umalen, scalar_importance = self.asr_model.uma(enc, enclen)
> 
> ```
> enc, umalen, scalar_importance = self.asr_model.uma(enc, enclen)
> enc, _ = self.asr_model.decoder(enc, umalen, torch.tensor(0), torch.tensor(0), self.asr_model.ctc)
> # Normal ASR
> if isinstance(enc, tuple):
>     enc = enc[0]
> 
> # obtain the tokens corresponding to the UMA output frames
> ys_hat = (self.asr_model.ctc.argmax(enc).data)[0].tolist()
> token = self.converter.ids2tokens(ys_hat)
> 
> valleys = scalar_importance.squeeze().cpu()
> valleys = valleys.tolist()
> utt_latency = {}
> for interval in utt_tg:
>     start_time = interval.minTime
>     end_time = interval.maxTime
>     text = interval.mark
>     if text in token:
>         # for simplicity, when there are repeated tokens, only the latency of the first occurrence will be counted.
>         index = token.index(text)
>         reg_time = (valleys[index]-1)*32
>         # if use lookahead: latency = reg_time - end_time*1000 + lookahead
>         latency = reg_time - end_time*1000
>            
>         if  start_time*1000 <= reg_time:
>             all_latency.append(latency)
>             utt_latency[text] = int(latency)
>             logging.info(f'latency: {text} {int(latency)}ms')
> ```
